dockstore-webservice/src/test/java/io/dockstore/webservice/languages/CWLHandlerPreprocessorTest.java

```diff
@@ -0,0 +1,151 @@
+        Assert.assertEquals(parse(V1_1 + WORKFLOW + "$mixin: v"), preprocess(V1_1 + WORKFLOW + "$mixin: v", set()));
+    }
+
+    @Test
+    public void testRun() {
+        final String runContent = "something: torun";
+        Assert.assertEquals(parse("run:\n  " + runContent), preprocess("run: b", set(file("/b", runContent))));
+        Assert.assertEquals(parse("run:\n  " + runContent), preprocess("run:\n  $import: b", set(file("/b", runContent))));
+    }
+
+    @Test
+    public void testMissingFile() {
+        Assert.assertEquals(Collections.emptyMap(), preprocess("$import: b", set()));
+        Assert.assertEquals("", preprocess("$include: b", set()));
+        Assert.assertEquals(parse(V1_0 + WORKFLOW + "a: x"), preprocess(V1_0 + WORKFLOW + "a: x\n$mixin: b", set()));
+    }
+
+    @Test
+    public void testMultilevelImports() {
+        final String imported = "levels: two";
+        Assert.assertEquals(parse(imported), preprocess("$import: b", set(file("/b", "$import: c"), file("/c", imported))));
+    }
+
+    @Test
+    public void testRelativeImport() {
+        final String imported = "some: content";
+        Assert.assertEquals(parse(imported), preprocess("$import: subsub/b", set(file("/sub/subsub/b", imported)), "/sub/a"));
+    }
+
+    @Test
+    public void testAbsoluteImport() {
+        final String imported = "some: content";
+        Assert.assertEquals(parse(imported), preprocess("$import: /b", set(file("/b", imported)), "/sub/a"));
+    }
+
+    @Test
+    public void testHttpUrlImport() {
+        Assert.assertEquals(Collections.emptyMap(), preprocess("$import: http://www.foo.com/bar", set()));
+        Assert.assertEquals(Collections.emptyMap(), preprocess("$import: https://www.foo.com/bar", set()));
+    }
+
+    @Test
+    public void testFileUrlImport() {
+        final String imported = "some: thing";
+        Assert.assertEquals(parse(imported), preprocess("$import: file://b", set(file("/b", imported))));
+    }
+
+    @Test
+    public void testRunOfNonexistentFile() {
+        final String runImport = "run:\n  $import: filename";
+        final String runReduced = "run: filename";
+        Assert.assertEquals(parse(runReduced), preprocess(runImport, set()));
+        Assert.assertEquals(parse(runReduced), preprocess(runReduced, set()));
+    }
+
+    @Test(expected = CustomWebApplicationException.class)
+    public void testMaxDepth() {
+        // preprocess a file that recursively imports itself
+        preprocess("$import: a", set(file("/a", "$import: a")));
+    }
+
+    private void preprocessManyIncludes(int includeCount, int includeSize) {
+        StringBuilder builder = new StringBuilder();
+        for (int i = 0; i < includeCount; i++) {
+            builder.append("a" + i + ":\n  $include: b\n");
+        }
+        preprocess(builder.toString(), set(file("/b", "x".repeat(includeSize))));
+    }
+
+    @Test(expected = CustomWebApplicationException.class)
```

Comment:
Should check for httpstatus and message if you forsee there ever being multiple messages in the future